First version of the sigsegv handler

zaregata opened the pull request:
Faut que je rajoute plus de test, je sais

Commit:
First version of the sigsegv handler

Commit:
Merge branch 'main' into implement_sigsegv_handler

Commit:
sigsev handler part 2

Commit:
sigsegv part3: everthing work edition

SamyB2 commented on src/library.c:
Mets les includes dans le .h directement, on laisse que comme seul include "library.h:

zaregata replied:
Alors c'est pas recommendé.
Parce que ça veut dire que lorsque quelqu'un include le header, il inclus aussi par transitivité toutes les librairies qui est seulement utiles en interne de la librairie.

et c'est mal. Un progamme ou partie de programme ce doit d'include le strict minimum et rien d'autre

SamyB2 replied:
Okok

SamyB2 commented on test/library/init_func_test.cpp:
un soucis avec les nombres qui se finiseent en 1 ???

zaregata replied:
C'est pour pouvoir lancer les deux tests à la suite.
J'ai eu des problèmes avant parce que c'était des ports déjà pris

SamyB2 replied:
Ça arrive quand des tests ne se sont pas bien terminé, tu peux faire un 
lsof -i :port afin de voir les proc qui les utilisent puis tu peux les kill

SamyB2 commented on test/sigsegv/sigsegv_test.cpp:
il fait rien ce test, on est d'accord ?

zaregata replied:
Je vais le supprimer, effectivement

SamyB2 commented on src/core/core.h:
essaie de merge main, vers ta branche actuelle car y a des conflits

SamyB2 commented on src/sigsegv_handler/sigsegv.c:
```diff
@@ -0,0 +1,111 @@
+    return dsm;
+}
+
+void exit_sigsegv(void * addr, size_t size) {
```
pas besoin de faire ça en vrai, vu que aucune mémoire a été allouée on peut juste quitter le système normalement

zaregata replied:
Faut bien enlever le handler quelque part nan ?

SamyB2 replied:
Ahh si, peut être que le programme ne finit pas quand on le user finit son utilisation
My bad...j'ai rien dit

SamyB2 commented on src/sigsegv_handler/sigsegv.c:
```diff
@@ -0,0 +1,111 @@
+}
+
+
+static void INVALIDATION_handler(struct message *message) {
```
tu ne modifie pas le tableau page_owners dans data_transfer.c avec la personne qui t'as send l'invalidation, car elle est la nouvelle détentrice de la page

zaregata replied:
Indeed

SamyB2 replied:
Tu as node_copy(dst, src) dans utils si jamais

Commit:
Merge branch 'main' into implement_sigsegv_handler

SamyB2 commented on src/sigsegv_handler/sigsegv.c:
```diff
@@ -0,0 +1,111 @@
+    // And the handler will be run once again... The circle of life. Beautiful.
+
+
+    size_t msg_size = sizeof(struct message) + sizeof(size_t);
```
ce code pourrait être mis dans une fonction à part, ce serait bien

zaregata replied:
Why not I guess

SamyB2 commented on src/sigsegv_handler/sigsegv.c:
```diff
@@ -0,0 +1,111 @@
+    size_t * page_id = (size_t *)(msg + 1);
+    *page_id = page_index;
+
+    broadcast_message(msg, msg_size);
```
pas besoin de broadcast si c'est un READ

zaregata replied:
On it

Commit:
Correcting the very true comment from the one and only SAMY

SamyB2 commented on src/sigsegv_handler/sigsegv.c:
```diff
@@ -0,0 +1,128 @@
+        exit(EXIT_FAILURE);
+    }
+
+    if (!is_owner) {
```
Tu peux prendre le nombre de pages en paramètres, je pense que dans les fonctions de init_DSM et join_DSM, le nombre de pages est set avant
Pour éviter de refaire les quelques calculs qui ont déjà été fait avant

bobbywan-kenoby commented on src/core/core.c:
pourquoi tu as des changement sur core.c ?
normalement ça ne devrait pas être possible...

bobbywan-kenoby commented on src/core/core.h:
ça me pose un problème les modifications d'interface comme ça...
si j'avais décider de gérer ma structure autrement ça aurait casser le module...
j'aimerais bien qu'on revois cette partie parce que je suis pas fan de ce lien qui est fait ici :thinking: potentiellement mettre le mode d'ouverture de la page a un autre endroit pour qu'il soit accessible pour toi

bobbywan-kenoby commented on src/network/network.c:
pourquoi les interfaces et le code d'autre module change ??

bobbywan-kenoby replied:
ok ça corrige des warnings qu'on avait dans le network :thinking: 
évitons quand même de modifier d'autre module que celui dev sinon j'en aurais bien chier si j'avais commencé a travailler sur l'opti du network avant la validation de ta pull request :sweat_smile:

bobbywan-kenoby commented on test/library/init_func_test.cpp:
si on pouvait éviter de faire ce genre de changement.... :joy:

SamyB2 commented on test/sigsegv/sigsegv_test.cpp:
Faut que tu écrive des tests pour vraiment tester le sigsegv

SamyB2 reviewed (changes requested):
Il faut juste que tu fasse des tests pour vérifier que le défaut de page marche de façon générale

Commit:
Merge branch 'main' into implement_sigsegv_handler

Commit:
Merge branch 'main' into implement_sigsegv_handler

Commit:
First working(?) version of the library

zaregata commented:
Pour les tests qui ne marche pas qu'est-ce que je fait, je les commentes en attendants que vous les corrigiez ?

J'essaie de les corriger moi-même ?

bobbywan-kenoby commented:
> Pour les tests qui ne marche pas qu'est-ce que je fait, je les commentes en attendants que vous les corrigiez ?
> 
> J'essaie de les corriger moi-même ?

c'est les tests de @SamyB2 qui crash ?
si c'est le cas vois avec lui pour les miens si yen a arrange les pour qu'il passe et dis moi ce que tu as fait on fera des ajustement si ya besoin dans la pull request

bobbywan-kenoby commented:
> Pour les tests qui ne marche pas qu'est-ce que je fait, je les commentes en attendants que vous les corrigiez ?
> 
> J'essaie de les corriger moi-même ?

ok c'est bon tu peux tous les adapter, on verra si ya des truc a changer jsuis chaud pour qu'on valide se soir la pull request !!

SamyB2 commented on src/sigsegv_handler/sigsegv.c:
```diff
@@ -0,0 +1,171 @@
+static void INVALIDATION_handler(struct message *msg) {
+    size_t *page_id = (size_t *) (msg + 1);
+    memory_lock(*page_id);
+    node_copy(page_owners + *page_id, &msg->sender);
```
Peux tu ne pas faire un node_copy directement, mais passe par une fonction qui te set le page owner, normalement dans la branche #26, tu peux reprendre le .h du data transfer (en ajoutant juste la fonction, set_new_owner ou un truc comme ça)

bobbywan-kenoby replied:
du coup il doit attendre que ta branche soit merge sur main ?? :scream:

SamyB2 commented on src/sigsegv_handler/sigsegv.c:
```diff
@@ -0,0 +1,171 @@
+    // And the handler will be run once again... The circle of life. Beautiful.
+
+    if (curr_writing && lock_status == WRITING) {
+        node_copy(page_owners + page_index, &me);
```
Pareil pour cette ligne, ne modifie pas page_owner directement, fais le à travers une fonction set
Je devrais rendre page_owner en static btw, vais le faire plus tard

bobbywan-kenoby commented on src/core/core.c:
```diff
@@ -169,6 +163,7 @@ static void handle_local_UNLOCK(int page_id, struct node_id *from)
 
 void unlock(size_t page_id, enum lock_type lock_type)
 {
+    memory_lock(page_id);
```
Tu n'as pas le droit de faire ça, il faudrait plutôt le faire dans library.c dans une fonction d'assemblage des modules comme lock_read, unlock_read, lock_write, unlock_write.

Commit:
Update test to work with the sigsegv handler

Commit:
Use the new API to change node's owner

bobbywan-kenoby commented on src/library.c:
```diff
@@ -43,9 +44,8 @@ static void INFO_DSM_handler(struct message *message)
 }
 
-static void set_all_handlers(void) 
+static void set_all_handlers() 
```
garde le void en paramètre c'est plus claire pour le compilateur il me semble

bobbywan-kenoby commented on src/network/network.c:
```diff
@@ -1,5 +1,5 @@
 #include "network.h"
-#include "library.h"
+#include "../library.h"
```
est ce que tu peux enlever cette include normalement on en n'a pas besoin

Commit:
Move memory lock and readd a void parameter

No idea how the void parameter disapeared honestly.
If it's indeed me who removed it, let it be known that I didn't do it on
purpose

bobbywan-kenoby reviewed (dismissed):
moyennant un petit nettoyage c'est ok pour moi

bobbywan-kenoby commented:
Mais ils sont passer en local les tests avant que tu push ?
Je commence a en douter vu le nombre de fois qu'on relance...

zaregata commented:
> Mais ils sont passer en local les tests avant que tu push ? Je commence a en douter vu le nombre de fois qu'on relance...

Ils passent, Ils passent.
J'ai même make clean et cmake -S . + exécuter les tests plusieurs fois pour être sur mais ils passent.

zaregata commented:
au pire, pull ma branche et test la sur ta machine

Commit:
Oops, forgot a print my b

SamyB2 commented on src/core/data_transfer.c:
```diff
@@ -1,16 +1,22 @@
     void *addr_np = (void *) (page_id + 1);
     void *addr_p = dsm + (*page_id) * PAGE_SIZE;
     node_copy(page_owners + *page_id, &message->sender);
+    memory_unlock_write(*page_id);
```
normalement il y a pas besoin e protéger cette partie là avec un mutex, car dans tous les cas il ne devrait pas avoir ni de read ni de write dessus, car les donnés ne sont pas à joure

SamyB2 commented on src/sigsegv_handler/sigsegv.c:
```diff
@@ -0,0 +1,171 @@
+            prot = PROT_NONE;
+            break;
+        case READING:
+        case WRITING:
```
pourquoi tu ne mets pas un PROT_WRITE pour le cas WRITING ?

SamyB2 commented on src/sigsegv_handler/sigsegv.c:
```diff
@@ -0,0 +1,171 @@
+    } else {
+        prot = PROT_NONE;
+    }
+    size_t nb_page = size / PAGE_SIZE;
```
tu peux te faire passer le nombre de page en argument ici, pas besoin de le re-calculer une 2eme fois

zaregata committed:
Merge branch 'main' into implement_sigsegv_handler

bobbywan-kenoby commented:
Il y a un autre problème.... le deadlock ne vient pas de la couche réseau j'ai rerun quelque fois les test dans la CI pour en être sûr....

zaregata committed:
Merge branch 'main' into implement_sigsegv_handler

SamyB2 reviewed (approved):
Enfin fini cette issue
